oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/DataModelService.kt

```diff
@@ -1,92 +1,176 @@
+    private fun updateIssueList() {
+        eventBus.post(IssuesUpdatedEvent(findIssue()))
+    }
+
+    fun setDuration(nodeName: String?, timeOut: Duration?) {
+        if (nodeName == null || nodeName !in proxyByName) {
             return
         }
-        val node: Simulation = simByName.getValue(nodeName)
+        val node = proxyByName.getValue(nodeName)
         val newNode = node.copy(timeOut = timeOut)
-        simByName += nodeName to newNode
+        proxyByName += nodeName to newNode
     }
 
-    override fun renameSim(target: Simulation, newName: String, oldName: String) {
-        require(simByName.containsKey(oldName) && !simByName.containsKey(newName))
-        simByName += newName to target
+    fun renameSim(newName: String, oldName: String): Boolean {
+        if (newName in simByName || oldName !in simByName) {
+            return false
+        }
+        simByName += newName to simByName.getValue(oldName).copy(name = newName)
         simByName -= oldName
-        removeNodeFromView(oldName)
-        addNodeToView(newName)
+        eventBus.post(PluginRenamedEvent(oldName, newName))
+        renameProxy(newName, oldName)
+        updateIssueList()
+        return true
     }
 
-    override fun removeSim(name: String) {
-        require(simByName.containsKey(name))
-        simByName -= name
-        removeNodeFromView(name)
+    private fun renameProxy(newName: String, oldName: String): Boolean {
+        if (newName in proxyByName || oldName !in proxyByName) {
+            return false
+        }
+        proxyByName += newName to proxyByName.getValue(oldName).copy(name = newName)
+        proxyByName -= oldName
+        eventBus.post(ProxyRenamedEvent(oldName, newName))
+        return true
     }
 
-    override fun addNewSim(simulation: Simulation) {
-        require( ! simByName.containsKey(simulation.name))
+    fun addNewSim(simulation: Simulation) {
+        require(simulation.name !in simByName)
         simByName += simulation.name to simulation
-        addNodeToView(simulation.name)
+
+        eventBus.post(PluginRegisteredEvent(simulation.name))
+        updateIssueList()
+    }
+
+    /**
+     * try unregister a node and remove it
+     */
+    fun closeSim(name: String) {
+        require(name in simByName)
+        try {
+            simByName.getValue(name).input.onCompleted()
+        } finally {
+            removeSim(name)
+        }
+    }
+
+    fun removeSim(name: String) {
+        require(name in simByName)
+        simByName -= name
+        eventBus.post(PluginUnRegisteredEvent(name))
+        updateIssueList()
     }
 
-    override fun updateSim(newNode: Simulation) {
-        require(simByName.containsKey(newNode.name))
+    fun updateSim(newNode: Simulation) {
+        require(newNode.name in simByName)
         simByName += newNode.name to newNode
-        rebindView()
+        eventBus.post(PluginUpdatedEvent())
+        updateIssueList()
```

Comment:
hmm, so we post an event but then update something and post another event?

Any chance we can get this down to a single event?

Reply:
Now interface ModelEvent will trigger problem finding to update